RecoTauTag/RecoTau/interface/DeepTauHelper.h

```diff
@@ -0,0 +1,1744 @@
+    inline auto getTauDxy(const pat::PackedCandidate& cand) { return cand.dxy(); }
+    inline auto getPvAssocationQuality(const reco::PFCandidate& cand) { return 0.7013f; }
+    inline auto getPvAssocationQuality(const pat::PackedCandidate& cand) { return cand.pvAssociationQuality(); }
+    inline auto getPuppiWeight(const reco::PFCandidate& cand, const float aod_value) { return aod_value; }
+    inline auto getPuppiWeight(const pat::PackedCandidate& cand, const float aod_value) { return cand.puppiWeight(); }
+    inline auto getPuppiWeightNoLep(const reco::PFCandidate& cand, const float aod_value) { return aod_value; }
+    inline auto getPuppiWeightNoLep(const pat::PackedCandidate& cand, const float aod_value) {
+      return cand.puppiWeightNoLep();
+    }
+    inline auto getLostInnerHits(const reco::PFCandidate& cand, float default_value) {
+      return cand.bestTrack() != nullptr
+                 ? cand.bestTrack()->hitPattern().numberOfLostHits(reco::HitPattern::MISSING_INNER_HITS)
+                 : default_value;
+    }
+    inline auto getLostInnerHits(const pat::PackedCandidate& cand, float default_value) { return cand.lostInnerHits(); }
+    inline auto getNumberOfPixelHits(const reco::PFCandidate& cand, float default_value) {
+      return cand.bestTrack() != nullptr
+                 ? cand.bestTrack()->hitPattern().numberOfLostHits(reco::HitPattern::MISSING_INNER_HITS)
+                 : default_value;
+    }
+    inline auto getNumberOfPixelHits(const pat::PackedCandidate& cand, float default_value) {
+      return cand.numberOfPixelHits();
+    }
+    inline auto getHasTrackDetails(const reco::PFCandidate& cand) { return cand.bestTrack() != nullptr; }
+    inline auto getHasTrackDetails(const pat::PackedCandidate& cand) { return cand.hasTrackDetails(); }
+    inline auto getPseudoTrack(const reco::PFCandidate& cand) { return *cand.bestTrack(); }
+    inline auto getPseudoTrack(const pat::PackedCandidate& cand) { return cand.pseudoTrack(); }
+    inline auto getFromPV(const reco::PFCandidate& cand) { return 0.9994f; }
+    inline auto getFromPV(const pat::PackedCandidate& cand) { return cand.fromPV(); }
+    inline auto getHCalFraction(const reco::PFCandidate& cand, bool disable_hcalFraction_workaround) {
+      return cand.rawHcalEnergy() / (cand.rawHcalEnergy() + cand.rawEcalEnergy());
+    }
+    inline auto getHCalFraction(const pat::PackedCandidate& cand, bool disable_hcalFraction_workaround) {
+      float hcal_fraction = 0.;
+      if (disable_hcalFraction_workaround) {
+        // CV: use consistent definition for pfCand_chHad_hcalFraction
+        //     in DeepTauId.cc code and in TauMLTools/Production/plugins/TauTupleProducer.cc
+        hcal_fraction = cand.hcalFraction();
+      } else {
+        // CV: backwards compatibility with DeepTau training v2p1 used during Run 2
+        if (cand.pdgId() == 1 || cand.pdgId() == 130) {
```

Comment:
I'm surprised this isn't checking `abs(cand.pdgId())`, but it does appear to be this way in official-cmssw. Maybe something else to check with the Tau POG.

Reply:
for neutral particles like pdgId==1 (HF candidates) and pdgId==130 (neutral hadrons), i think only positive values are available?

Reply:
Ah of course, I was thinking of actual PDG IDs rather than PF IDs.